kv/memberlist/dnsprovider.go

```diff
@@ -13,3 +15,31 @@ type DNSProvider interface {
 	// Addresses returns the latest addresses present in the DNSProvider.
 	Addresses() []string
 }
+
+type dnsProvider struct {
+	sync.Mutex
+	addr []string
+}
+
+func NewDNSProvider() DNSProvider { return &dnsProvider{} }
+
+func (d *dnsProvider) Resolve(ctx context.Context, addrs []string) error {
+	d.Lock()
+	defer d.Unlock()
+	for _, a := range addrs {
+		ips, err := net.LookupIP(a)
+		if err != nil {
+			return err
+		}
+		for _, ip := range ips {
+			d.addr = append(d.addr, ip.String())
+		}
+	}
+	return nil
+}
+
+func (d *dnsProvider) Addresses() []string {
+	d.Lock()
+	defer d.Unlock()
+	return d.addr
```

Comment:
I think you need to make a copy of this otherwise the underlying storage of the slice can still be modified after the caller starts using it (I could be wrong, we _do_ run tests with `-race` but this code feels like a race condition).